core/src/main/java/org/infinispan/security/impl/AuthorizationHelper.java

```diff
@@ -110,12 +113,24 @@ private boolean checkSubjectPermissionAndRole(Subject subject, AuthorizationConf
             subjectACL = computeSubjectACL(subject, configuration);
 
          int permissionMask = requiredPermission.getMask();
-         return subjectACL.matches(permissionMask) && (requestedRole != null ? subjectACL.containsRole(requestedRole) : true);
+         return subjectACL.matches(permissionMask) && checkSubjectRole(subjectACL, requestedRoles);
       } else {
          return false;
       }
    }
 
+   private boolean checkSubjectRole(SubjectACL subjectACL, Collection<String> requestedRoles) {
+      if (requestedRoles == null || requestedRoles.isEmpty()) {
+         return true;
+      }
+      for (String role : requestedRoles) {
```

Comment:
We really need to add support for role sorting at some point to speed up this check